### ISSUE TITLE
Filtering gives redundant node results in CreateNewNode Dialog

### ISSUE THREAD

**Issue by me2beats** [godotengine/godot#46427]
**Godot version:**

3.2.3, 3.2.4rc3

**OS/device including version:**

Linux

**Issue description:**

Filtering gives redundant node results in `CreateNewNode` Dialog

**Steps to reproduce:**
![3 2 4rc3](https://user-images.githubusercontent.com/16458555/109224226-7a674c80-77dd-11eb-8168-e541819c96f3.png)

**Minimal reproduction project:**

**Comment by Calinou** [comment]
cc @rcorre

**Comment by rcorre** [comment]
By "redundant" do you just mean "things that don't match the filter string"?
Is the intention of this dialog to "filter" or just to "search"? I'm not actually sure if it is working as intended or not.

**Comment by me2beats** [comment]
Yep there are results that don't match the filter string, like `KinematicBody2D`, `CollisionShape2D` when the string is `line`.
The expected results are `Line2D`, `LineEdit` etc that is results which have `line`.

**Comment by rcorre** [comment]
I think what is happening here is this:

1. A string matches if it contains the characters you typed in order (but other characters can be between them!)
2. A node matches if its parent node matches

Co`li`sio`n`Obj`e`ct2D matches, therefore its children like KinematicBody are all shown. I actually think both these behaviors are desireable. For example, I can find `KinematicBody` by typing `kbod`. Similarly, I can see all types of ranges by typing `Range`. What you're seeing is a slightly confusing side-effect of that, which I don't think is too annoying since the dialog typically focuses the best match (e.g. if I type `line<enter>` real fast I still get a `Line2D`).

**Comment by me2beats** [comment]
Ok, we see 2 features of node search here:
1. Co`li`sio`n`Obj`e`ct2D (when typing `line`) and
2. Searching in parent classes as well.

I think the main issue is the first one.
Imo this is how search shouldn't work, it's just annoying.
You enter fewer characters, but spend more time choosing from a large number of results.

As for (2) this sometimes makes sense, sometimes you may need it and sometimes not. So I think a proposal could be created like "Add checkbox named `Search in parent classes`.

**Comment by rcorre** [comment]
> Imo this is how search shouldn't work, it's just annoying.

I think that's highly subjective. Some folks probably think like you do, but others like me are really annoyed when we _do_ have to type in the characters in exact sequence.

Other good filtering UX I'm familiar with, such as [fzf](https://github.com/junegunn/fzf), [youcompleteme](http://ycm-core.github.io/YouCompleteMe/), and the blender F3 menu all behave like this. It is also consistent with other filtering UIs within Godot such as the quick-open-scene menu.

A good practical example of where this works well: lets say I want to add a `PopupMenu`.

- If I type `menu`, I get `MenuButton`
- If I type `pop`, I get `Popup`

With the scheme you propose, I'd have to type `PopupM`, but with the current UI I can just type `pm`.

In case all that sounded too critical -- I think it is good to challenge these UIs and make them better! Maybe we can improve this one somehow, but I don't want to lose the existing filtering behavior (or complicate it with a variety of different search behaviors you have to choose between).

Also, this might be a "proposal" rather than a bug, since this behavior goes all the way back to Godot2.

**Comment by akien-mga** [comment]
IMO this type of filtering works best if exact matches are given a higher precedence than fuzzy matches. So it's cool if `pm` matches `PopupMenu`, but `line` should definitely match `Line2D` or other nodes with the exact `line` substring first.

**Comment by rcorre** [comment]
To be clear, typing `line` _does_ focus `Line2D` first. It just happens to show other less-obvious matches around it.

**Comment by rcorre** [comment]
I suppose one option would be to work like [qutebrowser](https://www.qutebrowser.org/), which [treats whitespace as `.*`](https://github.com/rcorre/qutebrowser/blob/a7178298efe371be292d85ab0a2f1cf7943365e0/qutebrowser/completion/models/listcategory.py#L62-L66). So `line` would only match names that contain the characters "line" in sequence, but `k bod` would match `KinematicBody`. I would still prefer the current behavior, but this could be a middle-ground.

**Comment by akien-mga** [comment]
Ah ok, I was just confused by the screenshot where `KinematicBody2D` is highlighted. IMO the current behavior is fine as is, but we could maybe do a better job at conveying that it actually does subsequence matching. Maybe highlighting matched characters in results? And/or lowering opacity of non-exact matches?

**Comment by Zireael07** [comment]
+1 to lowering opacity of non-exact matches, that would handily solve the problem of "why does it show kinematicBody and not line2d" without getting rid of the utility that fuzzy matches offer to some.

**Comment by rcorre** [comment]
Opacity seems like a good solution. Another option is to highlight matched characters (screenshot from Qutebrowser):

![1614364105](https://user-images.githubusercontent.com/2496231/109340120-96e3a300-7836-11eb-8cef-4d355a79bc51.png)

that's a little more work (since you have to pull out regex groups) but it does make it obvious _why_ something matched.

**Comment by hinlopen** [comment]
@akien-mga @rcorre This issue is already fixed in master with #40033 by implementing search ranking. Highlighting matched characters is fine, but somewhat unnecessary as you can almost always select the result the user wants after typing 2-3 characters with proper ranking.

**Comment by hinlopen** [comment]
What remains confusing is that the selected item will jump around in the search results window as you type new characters. I have found that centering the selected item in the middle of the result window helps a lot (see #40140), especially if you type fast.

More on topic: The issue here is not about filtering, but about ranking. With exact filtering you will still get into situations like this, just less often. It will also prevent people from searching for something like "kin3D" to get a KinematicBody3D node without typing the full length name.

Perhaps it would be good to have add a Quick Create New Node dialog akin to Quick Open with the same results as the regular dialog, but without building up the inheritance tree for each result.

**Comment by MJBrune** [comment]
Hmm so the desired result of ```pm``` finding ```PopupMenu``` and the desired result of finding ```Line2D``` when searching ```line``` are still very different than the result we are getting which has ```line``` found in CollisionObject. It might be useful to only focus on searching, not filtering, then add word separation to the searching. E.g. ```pm``` would work because ```PopupMenu``` starts with ```P``` and ```M``` words (capitalization.) but it wouldn't incorrectly find ```CollisionObject``` for ```line``` but it would pop up with ```CO``` because **C**ollision**O**bject.

Just a thought. This way feels like the best of both worlds while trying to ignore garbage results. I don't think anyone looks at CollisionObject and things "line" but they think ColObj or CoOb or CO. So the focus might be better on thinking what the user is looking for and displaying it like search rather than a pure filter of if these words contain these letters in this order.

**Comment by hinlopen** [comment]
That sounds fine to me. That should lower the amount of spurious results significantly, and with the inheritance tree every result adds a lot of visual complexity. I think the benefits outweigh the few instances where these results might be useful. I will take a stab at implementing this soon.

**Comment by rcorre** [comment]
> E.g. pm would work because PopupMenu starts with P and M words (capitalization.)

[YouCompleteMe](http://ycm-core.github.io/YouCompleteMe/#completion-string-ranking) implemented this sort of ranking very well, you might peek at their code for reference. It might be more complicated than what we need here, but as a user it felt like magic :)

**Comment by KoBeWi** [comment]
Looks like it's fixed?

<img width="768" height="500" alt="Image" src="https://github.com/user-attachments/assets/7dbd2611-8ebe-40ff-8aa1-786edfec56b6" />
